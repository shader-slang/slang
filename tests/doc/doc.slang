//TEST:DOC:-entry computeMain -target hlsl -stage compute -doc -no-codegen

void outFunc(out int v)
{
    v = 10;
}

/// Testing out nested generics
struct ParentStruct<T> ///< Some type
{
    /// Testing out a child
    struct ChildStruct<S> ///< Some other type
    {
        /// A useless method hey ho
        T getValue(S v) { T t; S s; return t; }
    };
};

struct GenericStruct<T>
{
        /// Let's try a typedef too
    typedef T Element;

    T getValue() { return value; }
    
    T value;
};

/// A rather silly generic function to test out doc extraction
T addInts<T : __BuiltinIntegerType,     ///< The type we are operating on 
    /// Just testing out a 
    /// non type based generic
    let U : int,
    let V : int>    ///< And another one                       
    (
    /// CHECKING!!
    T z,           ///< The Z parameter
    T b)            ///< The B parameter
{
    return z + b;
}

namespace Hey
{
    void doAnotherThing(int a);
}

    /// Let's test indent
    /// 
    /// ```
    /// {
    ///     imIndented();
    /// }   
    /// ```
    ///
RWStructuredBuffer<int> inputBuffer;       


/// An interface to do things
interface IDoThing
{
        /// An associated type
    associatedtype V;
    
        /// Add two integers
    V add(V a,      ///< First parameter 
        V b           ///< Second parameter
    );
    
        /// Subtract
        /// Multi-line
    int sub(int a,      ///< First
            int b       ///< Second
    );
}

interface IThing
{
    float getValue();
};

/// Implement IThing on float
extension float : IThing
{
        /// Just return the float itself!
	float getValue() { return this; }
}

struct Thing : IThing, IDoThing
{
    typedef int V;
    
    int add(int a, int b) { return a + b; }
    int sub(int a, int b ) { return a - b; }
    float getValue() { return 1.0f; }
};

/// A struct with some fields
struct SomeStruct
{
    /// A field
    int aField;
    /// Multi-line 
    /// is a thing
    int anotherField;
    int yetAnother; ///< A field with stuff 
    
        /// Get a value 
    int getMethod() { return yetAnother; }
};

/// An enum
enum AnEnum
{
    Value, ///< A value
    /// Another value
    /// With a multi-line comment
    AnotherValue,
};


//TEST_INPUT:ubuffer(data=[0 0 0 0], stride=4):out,name outputBuffer
RWStructuredBuffer<int> outputBuffer;       ///< An output buffer

/// doThing!
int doThing(int a,  ///< a parameter
            int b)  ///< b parameter
{
    while (b >= 0)
    {
        a 
        += 
        a;
    }
    
    return a;
}


[numthreads(4, 1, 1)]
void computeMain(uint3 dispatchThreadID : SV_DispatchThreadID)
{
    int a = dispatchThreadID.x;
    int b = dispatchThreadID.y;
    int c = dispatchThreadID.z;
    int d = a + b * c;
    int e = d + c / 2;
    
    for (int i = 0; i < b; ++i)
    {
        if (e > 10 && i & 2)
        {
            a += b; b -= c; c += c; d = d + e + a; e = a;
        }
        else
        {
            a = e; b = c + c; d += d + __SyntaxError(); e = doThing(e, dispatchThreadID.x);
            
        }
    }
    
    outputBuffer[dispatchThreadID.x] = a + b + c + d + e;
}
