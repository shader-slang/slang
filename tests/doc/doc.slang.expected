result code = 0
standard error = {
--------------------------------------------------------------------------------
# `outFunc`

## Signature 

```
void outFunc(out int v);
```

## Parameters

* `v`

--------------------------------------------------------------------------------
# `struct ParentStruct<T>`

## Description

Testing out nested generics

## Generic Parameters

* `T` Some type 

--------------------------------------------------------------------------------
# `struct ParentStruct<T>.ChildStruct<S>`

## Description

Testing out a child

## Generic Parameters

* `S` Some other type 

## Methods

* `getValue`

--------------------------------------------------------------------------------
# `ParentStruct<T>.ChildStruct<S>.getValue`

## Description

A useless method hey ho

## Signature 

```
T ParentStruct<T>.ChildStruct<S>.getValue(S v);
```

## Parameters

* `v`

--------------------------------------------------------------------------------
# `struct GenericStruct<T>`

## Generic Parameters

* `T`

## Fields

* `value`

## Methods

* `getValue`

--------------------------------------------------------------------------------
# `GenericStruct<T>.getValue`

## Signature 

```
T GenericStruct<T>.getValue();
```

--------------------------------------------------------------------------------
# `addInts`

## Description

A rather silly generic function to test out doc extraction

## Signature 

```
T addInts<T, U:int, V:int>(
    T                    z,
    T                    b);
```

## Parameters

* `T` The type we are operating on  
* `U` Just testing out a  non type based generic 
* `V` And another one                        
* `z` CHECKING!! 
* `b` The B parameter 

--------------------------------------------------------------------------------
# `Hey::doAnotherThing`

## Signature 

```
void Hey::doAnotherThing(int a);
```

## Parameters

* `a`

--------------------------------------------------------------------------------
# inputBuffer

```
RWStructuredBuffer inputBuffer
```

## Description

Let's test indent

```
{
    imIndented();
}   
```

--------------------------------------------------------------------------------
# `interface IDoThing`

## Description

An interface to do things

# Associated types

* _V_ An associated type 


## Methods

* `add`
* `sub`

--------------------------------------------------------------------------------
# `IDoThing.add`

## Description

Add two integers

## Signature 

```
V IDoThing.add(
    V                    a,
    V                    b);
```

## Parameters

* `a` First parameter  
* `b` Second parameter 

--------------------------------------------------------------------------------
# `IDoThing.sub`

## Description

Subtract
Multi-line

## Signature 

```
int IDoThing.sub(
    int                  a,
    int                  b);
```

## Parameters

* `a` First 
* `b` Second 

--------------------------------------------------------------------------------
# `interface IThing`

## Methods

* `getValue`

--------------------------------------------------------------------------------
# `IThing.getValue`

## Signature 

```
float IThing.getValue();
```

--------------------------------------------------------------------------------
# `extension float : IThing`

*Implements:* `IThing`

## Description

Implement IThing on float

## Methods

* `getValue`

--------------------------------------------------------------------------------
# `float.getValue`

## Description

Just return the float itself!

## Signature 

```
float float.getValue();
```

--------------------------------------------------------------------------------
# `struct Thing`

*Implements:* `IThing`, `IDoThing`

## Methods

* `add`
* `sub`
* `getValue`

--------------------------------------------------------------------------------
# `Thing.add`

## Signature 

```
int Thing.add(
    int                  a,
    int                  b);
```

## Parameters

* `a`
* `b`

--------------------------------------------------------------------------------
# `Thing.sub`

## Signature 

```
int Thing.sub(
    int                  a,
    int                  b);
```

## Parameters

* `a`
* `b`

--------------------------------------------------------------------------------
# `Thing.getValue`

## Signature 

```
float Thing.getValue();
```

--------------------------------------------------------------------------------
# `struct SomeStruct`

## Description

A struct with some fields

## Fields

* `aField` A field 
* `anotherField` Multi-line  is a thing 
* `yetAnother` A field with stuff  

## Methods

* `getMethod`

--------------------------------------------------------------------------------
# `SomeStruct.getMethod`

## Description

Get a value 

## Signature 

```
int SomeStruct.getMethod();
```

--------------------------------------------------------------------------------
# enum AnEnum

## Values 

* _Value_ A value 
* _AnotherValue_ Another value With a multi-line comment 
## Description

An enum

--------------------------------------------------------------------------------
# outputBuffer

```
RWStructuredBuffer outputBuffer
```

## Description

An output buffer

--------------------------------------------------------------------------------
# `doThing`

## Description

doThing!

## Signature 

```
int doThing(
    int                  a,
    int                  b);
```

## Parameters

* `a` a parameter 
* `b` b parameter 

--------------------------------------------------------------------------------
# `computeMain`

## Signature 

```
void computeMain(vector<uint,3> dispatchThreadID);
```

## Parameters

* `dispatchThreadID`

}
standard output = {
}
